# Lilac

## Introduction
TODO - copy across documentation from the Lilac Google doc


## Configuration
To install, download the latest compiled jar file from the [download links](#version-history-and-download-links). 

### Resource Files
Lilac requires the following resource files:
- Ref Genome FASTA
- HLA allele nucleotide and amino acid definitions, generated by Lilac (see below)

Reference files are available here [HMFTools-Resources](https://resources.hartwigmedicalfoundation.nl/):

#### Generate HLA Definitions
Nucleotide and amino acid allele defintions are available here:
- https://www.ebi.ac.uk/ipd/imgt/hla/download.html
- ftp://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/


### Mandatory Arguments

Argument | Description 
---|---
sample | Sample ID
reference_bam | Sample germline BAM

### Optional Inputs

Argument | Description 
---|---
tumor_bam | Sample tumor BAM
gene_copy_number | Sample gene copy number file from Purple
expected_alleles | List of alleless separated by ';'. These alleles will have their coverage and ranking reported even if not in the winning solution
restricted_alleles | List of alleless separated by ';'. Restrict evaluation to these alleles only.

### Optional Arguments

Argument | Default | Description 
---|---|---
min_base_qual|30| Min base quality for BAM reads, see documentation for details 
min_evidence|2|Min evidence in terms of supporting fragments (1 high-quality, 1 other) 
min_fragments_per_allele|7|See documentation for details 
min_fragments_to_remove_single|40|See documentation for details 
top_score_threshold|5|Write results for solutions lower than the top solution by this difference in score 
threads|1|Number of threads to use

NOTE: Lilac handles BAMs which have been sliced for the HLA gene regions.

### Example Usage

```
java -jar lilac.jar \
   -sample COLO829T 
   -reference_bam /sample_data_path/COLO829R.bam \
   -ref_genome /path_to_ref_genome_fasta_file/ \
   -resource_dir /path_to_lilac_resource_files/ \ 
   -output_dir /output_dir/ \
   -threads 16 \
```

Or with tumor BAM, somatic variants VCF and Purple gene copy number inputs:

```
java -jar lilac.jar \
   -sample COLO829T 
   -reference_bam /sample_data_path/COLO829R.bam \
   -tumor_bam /sample_data_path/COLO829T.bam \
   -somatic_vcf /sample_data_path/COLO829T.purple.somatic.vcf.gz \
   -gene_copy_number /sample_data_path/COLO829T.purple.cnv.gene.tsv \
   -ref_genome /path_to_ref_genome_fasta_file/ \
   -resource_dir /path_to_lilac_resource_files/ \ 
   -output_dir /output_dir/ \
   -threads 16 \
```

## Output

The following tab delimited files are written:

File | Description
--- | ---
SAMPLE_ID.HLA-A.aminoacids.txt|Fragment support for each amino acid by HLA gene
SAMPLE_ID.HLA-A.nucleotides.txt|Fragment support for each nucleotide by HLA gene
SAMPLE_ID.candidates.aminoacids.txt|Fragment support for amino acids in the candidate alleles
SAMPLE_ID.candidates.nucleotides.txt|Fragment support for nucleotides in the candidate alleles
SAMPLE_ID.lilac.qc.tsv|QC metrics
SAMPLE_ID.lilac.tsv|Summary coverage for the HLA gene solution


# Version History and Download Links
- [1.0](https://github.com/hartwigmedical/hmftools/releases/tag/lilac-v1.0)
 